### PR TITLE
Policy refactoring with automation and testing

### PR DESCRIPTION
**Pull request by krowlandson**
## Overview/Summary

This PR makes a fundamental change to the how we manage and test custom policies implemented as part of Azure landing zones.

## This PR fixes/adds/changes/removes

1. Adds a new resource library containing individual policy definition and policy set definition resources as individual files
2. Adds a Bicep template used to programmatically generate a new `policies.json` file for the `eslzArm` Portal deployment
3. The new `policies.json` template is now designed to universally work across `AzureCloud`, `AzureChinaCloud` and `AzureUsGovernment` clouds
4. Adds new GitHub Action to automatically regenerate `policies.json` when relevant changes are detected in a PR
5. Adds new GitHub Action to perform static code analysis (linting) of the new `src` folder (also scans `eslzArm` folder but for reporting only)
6. Adds a new GitHub Action to perform automated testing of the `eslzArm` deployment, including a complete deployment and tear-down
7. Adds an updated version of the original `EnterpriseScaleLibraryTools` PowerShell module from the Terraform implementation, now rebranded as `Alz.Tools` and extended to include new functionality needed for broader use (_will be re-usable for Terraform and Bicep implementations_)
8. Adds a new GitHub Action to keep the `Alz.Tools` module up to date with the latest API versions
9. Includes minor (primarily cosmetic) updates to the `eslzArm` deployment to improve maintainability and control whilst running programmatically
10. Update elements in Portal UI definition to match the parameters set by their output for easier maintenance
11. Added logic to customize the Portal UI questions based on target Cloud environment, to improve support for `AzureChinaCloud` (_pending update to MG scope deployment_) and `AzureUSGovernment`
12. Improve consistency across deployment names and update to use `alz-` prefix instead of `EntScale-`
13. Update description field in `emailContactAsc` parameter across Cloud environments for `Deploy-MDFC-Config` Policy Assignment

### Breaking Changes

None identified

### `AzureCloud`
See test pipelines in GitHub Actions.

### `policies.json` for `AzureChinaCloud`

image

image

### `policies.json` for `AzureUsGovernment`

MicrosoftTeams-image

MicrosoftTeams-image (1)

### Testing URLs

n/a

## As part of this Pull Request I have

- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/Enterprise-Scale/pulls)
- [x] Associated it with relevant [issues](https://github.com/Azure/Enterprise-Scale/issues), for tracking and closure.
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/Enterprise-Scale/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated relevant and associated documentation.
- [x] Updated the ["What's New?"](https://github.com/Azure/Enterprise-Scale/wiki/Whats-new) wiki page (located: `/docs/wiki/whats-new.md`)

### CODE REVIEW

**Comment by faister** [comment]
@krowlandson I'll have a go with testing this in Azure China, will reach out when I have questions

**Comment by matt-FFFFFF** [comment]
I have one concern that we are using a Bicep data structure to store the configuration for the public/mc/ff policies.

This should be in a common data format like JSON or YAML.

I understand there is a feature request in the Bicep team for a new function (`loadJsonContent()`) but they have not committed to producing this. As a workaround to this I would like us to store the data as JSON/YAML, then convert this to Bicep in the absence of a `loadJsonContent()` func. This way we can start on the journey to have a machine readable definition for ALZ.

File: .github/workflows/test-portal.yml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+          creds: '{"clientId":"${{ secrets.AZURE_CLIENT_ID }}","clientSecret":"${{ secrets.AZURE_CLIENT_SECRET }}","subscriptionId":"${{ secrets.AZURE_SUBSCRIPTION_ID }}","tenantId":"${{ secrets.AZURE_TENANT_ID }}"}'
+          enable-AzPSSession: true
+        env:
+          AZURE_CLIENT_SECRET: ${{ secrets.AZURE_CLIENT_SECRET }}
+
+      - name: Register subscriptions
+        uses: azure/powershell@v1
+        with:
+          inlineScript: |
+            ./src/scripts/Invoke-ActionRegisterSubscriptions.ps1
```
Comment (jfaurskov) [inline .github/workflows/test-portal.yml:78]: Do we not run the risk of collisions with this approach, i.e. if more PRs are approved independently? Would it make more sense to generate new subscriptions for each deployment and trash after PR close?

**Review by jfaurskov — APPROVED** [review]
Other than the comments previously made as well as the single comment on collisions I have no further input. Sterling work @krowlandson !

**Review by jtracey93 — APPROVED** [review]

**Review by matt-FFFFFF — APPROVED** [review]
Great work

File: .github/workflows/test-portal.yml (hunk 1 above)
Comment (krowlandson) [reply .github/workflows/test-portal.yml:78]: Good question. This is a matter of choice ultimately and I prefer the idea that we co-ordinate runs for PRs, keeping in mind the only "collision" per-say is the placement of Subscriptions in a Management Group. All other resources (and the corresponding clean-up activities) are run on the basis of the unique `rootId` provided under each PR so we could theoretically run multiple PRs in parallel without too many issues.

However I don't consider this to be a likely event given current rate of change, and IMHO that downside far outweighs the issues associated with a high churn rate of Subscription creation and deletion operations needed to ensure unique Subscriptions per PR.

**Pull request merged**